src/approved-by.ts

```diff
@@ -0,0 +1,117 @@
+import * as core from "@actions/core";
+import * as github from "@actions/github";
+import { components } from "@octokit/openapi-types";
+import { GitHub } from "@actions/github/lib/utils";
+
+export type Octokit = InstanceType<typeof GitHub>;
+export type Review = components["schemas"]["pull-request-review"];
+export type Reviews = Review[];
+export type Reviewer = {
+  username: string;
+  name: string;
+};
+export type Reviewers = Reviewer[];
+
+export function getApprovedReviews(reviews: Reviews): Reviews {
+  const approved: Reviews = [];
+
+  const latestReviews = reviews
+    .reverse()
+    .filter((review) => review.state.toLowerCase() !== "commented")
+    .filter((review, index, array) => {
+      // https://dev.to/kannndev/filter-an-array-for-unique-values-in-javascript-1ion
+      return array.findIndex((x) => review.user?.id === x.user?.id) === index;
+    });
+
+  for (const review of latestReviews) {
+    core.debug(`Latest ${review.user?.login} review '${review.state.toLowerCase()}'`);
+
+    if (review.state.toLowerCase() === "approved") {
+      approved.push(review);
+    }
+  }
+
+  return approved;
```

Comment:
Do we really need so many loops here(reverse, filter, filter+findIndex, for of)? I might be missing that this is some specific order that needs to be fulfilled. But if not, wouldn't something like this be the same:

```
const approved = [];

loop reviews
  if (!reviewCommented && reviewApproved && !approved.find(approved.user.id === review.user.id)) {
    approved.push(review);
  }
  
return approved;  
```

Reply:
Reviews are returned in 
* [chronological order](https://docs.github.com/en/rest/pulls/reviews?apiVersion=2022-11-28#list-reviews-for-a-pull-request)
* it can return multiple reviews or declines from the same user

We only need the latest state.